### DISCUSSION TITLE
How to save/load a model

### DISCUSSION THREAD

**Discussion by daviddralle** [thodson-usgs/ratingcurve#67] [Q&A]
I am hoping to use this package to manage a large number of rating curves for a multi-catchment project. It would be really helpful to be able to save then reload a model, but I'm not certain the syntax for doing this with the existing class methods. Can you all provide a short example of how to save then reload a power law rating model? Thanks for any help. 

For reference, I have tried the following. Assume `powerrating` is a 3 segment `PowerLawRating` fitted to some data. Here, I try to save it (as netcdf?) then reload it in `newmodel`:

```
powerrating.save('./test.nc4')
newmodel = PowerLawRating(segments=3)
newmodel.load('./test.nc4')
```

**Comment by thodson-usgs** [comment]
Adding an example to the getting started notebook, but I'll suggest a slight modification.
```python
powerrating.save('./green_rating_fit.nc')
loaded_rating = PowerLawRating().load('./green_rating_fit.nc')
```

I think your example will also work. Did you try it?

**Comment by daviddralle** [comment]
Thanks for the quick reply! Here's what I tried: 

```python
powerrating.save('./fit.nc')
loaded_rating = PowerLawRating().load('./fit.nc')
```
and got the error: 
```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
[<ipython-input-15-20b3ad753162>](https://localhost:8080/#) in <cell line: 2>()
      1 powerrating.save('./fit.nc')
----> 2 loaded_rating = PowerLawRating().load('./fit.nc')

[/usr/local/lib/python3.10/dist-packages/ratingcurve/ratingmodel_builder.py](https://localhost:8080/#) in load(cls, fname)
    286         # All previously used data is in idata.
    287         if model.id != idata.attrs["id"]:
--> 288             raise ValueError(
    289                 f"The file '{fname}' does not contain an inference data of "
    290                 "the same model or configuration as '{cls._model_type}'"

ValueError: The file './fit.nc' does not contain an inference data of the same model or configuration as '{cls._model_type}'
```

Do i need to somehow save the idata separately and load that in?

**Reply by thodson-usgs** [reply]
No, I don't think so, but save/load hasn't been thoroughly tested.
I'm a bit busy, but I'll try to replicate this error in Python 3.10. It worked fine when I tried on 3.11 (12?).

PyMC is always changing too, so you might try updating your environment, or creating a new environment on 3.11+.

Keep me posted.

PS. Standard practice at USGS is not to save the model and fit. Rather we save the rating table (`powerrating.table()`) and use that to generate data. But I agree that it'd be nice to be able to feed a stage timeseries into a saved model.

**Reply by daviddralle** [reply]
It looks to me like the "segments" argument in the PowerLawRating() is messing with my save, specifically, I was using segments=3. When I switch to `segments=2`, everything works. 

And thanks for the low-down on USGS standard operating procedure! I mostly wanted to save the parameterization in case some reviewer down the line asks for the parameters of the power law rating equations. Generally though, I prefer the table approach also as you don't have to worry about saving/loading complex parameterizations. 

Thanks for your help and time!

**Reply by LZ-Eyasco** [reply]
Hi thodson, 

I meet the same issue like daviddralle met. 

The sample program works fine with segment = 2, but the error shows up after I tried different segment. 

I am wondering do I need to put any parameters into the load instruction if I am using segment = 3 or other values? Thanks

**Reply by thodson-usgs** [reply]
Could you post a code snippet that reproduces the error and I'll take a look.
